test/e2e/backup_restore_test.go

```diff
@@ -0,0 +1,231 @@
+
+		// 7. Ensure that all data is in the target cluster
+		assertNumEntries(cacheName, targetCluster+"-0", numEntries, cluster.Client)
+	}
+}
+
+func waitForValidBackupPhase(name, namespace string, phase v2.BackupPhase) {
+	err := wait.Poll(10*time.Millisecond, tconst.TestTimeout, func() (bool, error) {
+		backup := testKube.GetBackup(name, namespace)
+		if backup.Status.Phase == v2.BackupFailed && phase != v2.BackupFailed {
+			return true, errors.New("Backup failed")
+		}
+		return phase == backup.Status.Phase, nil
+	})
+	tutils.ExpectNoError(err)
+}
+
+func waitForValidRestorePhase(name, namespace string, phase v2.RestorePhase) {
+	err := wait.Poll(10*time.Millisecond, tconst.TestTimeout, func() (bool, error) {
+		restore := testKube.GetRestore(name, namespace)
+		if restore.Status.Phase == v2.RestoreFailed {
+			return true, errors.New("Restore failed")
+		}
+		return phase == restore.Status.Phase, nil
+	})
+	tutils.ExpectNoError(err)
+}
+
+func datagridService(name, namespace string, replicas int) *v1.Infinispan {
+	infinispan := cacheService(name, namespace, replicas)
+	infinispan.Spec.Service = v1.InfinispanServiceSpec{
+		Type: v1.ServiceTypeDataGrid,
+	}
+	return infinispan
+}
+
+func cacheService(name, namespace string, replicas int) *v1.Infinispan {
+	return &v1.Infinispan{
+		TypeMeta: tconst.InfinispanTypeMeta,
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+		Spec: v1.InfinispanSpec{
+			Replicas: int32(replicas),
+		},
+	}
+}
+
+func populateCache(cacheName, pod string, numEntries int, client ispnclient.HttpClient) {
+	headers := map[string]string{"Content-Type": "application/json"}
+
+	post := func(url, payload string, status int) {
+		rsp, err, _ := client.Post(pod, url, payload, headers)
+		tutils.ExpectNoError(err)
+		if rsp.StatusCode != status {
+			panic(fmt.Sprintf("Unexpected response code %d", rsp.StatusCode))
+		}
+	}
+
+	url := fmt.Sprintf("/rest/v2/caches/%s", cacheName)
+	config := fmt.Sprintf("{\"distributed-cache\":{\"name\":\"%s\"}}", cacheName)
+	post(url, config, http.StatusOK)
+
+	for i := 0; i < numEntries; i++ {
+		url = fmt.Sprintf("/rest/v2/caches/%s/%d", cacheName, i)
+		value := fmt.Sprintf("{\"value\":\"%d\"}", i)
+		post(url, value, http.StatusNoContent)
+	}
+}
+
+func assertNumEntries(cacheName, pod string, numEntries int, client ispnclient.HttpClient) {
+	url := fmt.Sprintf("/rest/v2/caches/%s?action=size", cacheName)
+	rsp, err, _ := client.Get(pod, url, nil)
+
+	tutils.ExpectNoError(err)
+	if rsp.StatusCode != http.StatusOK {
+		panic(fmt.Sprintf("Unexpected response code %d", rsp.StatusCode))
+	}
+}
+
+// Make sure that both the Infinispan cluster and zero pod exist and are ready
+func waitForZeroPodCluster(name, namespace string, clusterSize int, cluster *ispn.Cluster) {
+	zeroClusterSize := clusterSize + 1
+	testKube.WaitForPods(zeroClusterSize, tconst.SinglePodTimeout, &client.ListOptions{Namespace: namespace},
+		func(pods []corev1.Pod) bool {
```

Comment:
@ryanemerson we can avoid pooling for all pods and looking for `p.Name == name` just adding specific label to zero capacity pod and using a label selector for it

Reply:
This method is actually waiting to ensure that the pod exists and it has successfully joined the cluster.

Reply:
Yes, true, but we can easily use label selectors for rather than comparing by name

Reply:
I tried updating to the following, but no luck. I'll take a look later if I get a chance.

```
// Make sure that both the Infinispan cluster and zero pod exist and are ready
func waitForZeroPodCluster(name, namespace string, clusterSize int, podLabels map[string]string, cluster *ispn.Cluster) {
	testKube.WaitForPods(1, tconst.SinglePodTimeout, &client.ListOptions{
		Namespace:     namespace,
		LabelSelector: labels.SelectorFromSet(podLabels),
	}, nil)
	wait.Poll(tconst.DefaultPollPeriod, tconst.SinglePodTimeout, func() (done bool, err error) {
		return AssertClusterSize(clusterSize+1, name, cluster)
	})
}
```

Reply:
We should avoid using CreateOrUpdate in case if we just need to create it only and nothing to update. It's very easy operation. Pod for it's nature very limited and short life structure.